Rsyslog files permissions fix wildcard path

lonicerae opened the pull request:
* A simple test case as follows:

/etc/rsyslog.conf
```
include(file="/etc/rsyslog.d/*/*.conf" mode="optional")
include(file="/etc/rsyslog.d/*.conf" mode="optional")
```

/etc/rsyslog.d/subdir/custom1.conf
```
local1.* /tmp/local1.out
```

/etc/rsyslog.d/custom2.conf
```
local2.* /tmp/local2.out
```

* Rationale:

With current code, the RSYSLOG_INCLUDE or RSYSLOG_INCLUDE_CONFIG will be read as a string '/etc/rsyslog.d/*.conf' in the for loop.
This patch ensures the wildcard path is expanded to suitable files and then added to the array.

lonicerae commented:
Hello. This PR contains the identical fix to the old one https://github.com/ComplianceAsCode/content/pull/8543 .Test files have been added, please check. Thanks.

lonicerae commented:
/ok-to-test

Mab879 commented:
/ok-to-test

yuumasato commented:
Ping @matejak for his Bash skills, :)

matejak commented on linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/tests/include_config_syntax_perms_0600.pass.sh:
```diff
@@ -0,0 +1,56 @@
+
+source $SHARED/rsyslog_log_utils.sh
+
+PERMS=0600
+
+# setup test data
+create_rsyslog_test_logs 3
+
+# setup test log files and permissions
+chmod $PERMS ${RSYSLOG_TEST_LOGS[0]}
+chmod $PERMS ${RSYSLOG_TEST_LOGS[1]}
+chmod $PERMS ${RSYSLOG_TEST_LOGS[2]}
+
+# create test configuration file
+conf_subdir=${RSYSLOG_TEST_DIR}/subdir
+mkdir ${conf_subdir}
+test_subdir_conf=${conf_subdir}/test_subdir.conf
+test_conf=${RSYSLOG_TEST_DIR}/test.conf
```
Would it be possible to extract this to a function and put it to `rsyslog_log_utils.sh`? I think that we have a copy-pasted code here.

lonicerae replied:
Hi @matejak , do you mean the `cat << EOF ...` portion? I think so. Maybe it's better to leave this work to another commit for enhancement as I see some other files also include similar code and we may do a refactor on all of them. How do you think?

yuumasato commented:
@brett060102 Hi, you recently made changes to this remediation. Would you have any comments?

matejak commented:
Generally, I don't see anything obviously wrong or suspicious in the Bash code, so I would stick to the general recommendation of relying on test scenarios.

brett060102 commented:
Not an issue for this review, but couple flavors of the following are supported.
```
include(
   file=`echo $ENV_VAR`
   mode="optional"
)
''''
Which I don't find amusing.

And should have check during previous comment but directories are supported via:
$IncludeConfig /etc/rsyslog.d/

We might just want to drop the depth limit on the find. But still not sure oval will handle that and I am not an expert there.

brett060102 commented:
I think that the find should stay. Mode optional in following means that the no error generated if file not present.
```
include(
   file="/path/to/include.conf"
   mode="optional"
)
```
So, trying to remediated that file could result on an error in remediation. Depending on how it is run, it could terminate remediation which would mean that later remediations are not done.  I think this would only effect running the generated build/bash scripts directly and then only with "set -e", but we were doing exactly that.

lonicerae commented:
Hi @brett060102 , thanks for your comment. In fact I've submitted the code a month ago, however in the old PR it's not accepted because I did not create the testing code in time. I did a comparison here https://github.com/ComplianceAsCode/content/pull/8543#issuecomment-1116458344 , maybe this can help to understand the differences.

lonicerae commented:
Hi @yuumasato , thanks for your review. Please help to check the latest one which I have adjusted. Thank you.

brett060102 commented:
@lonicerae I had look over your initial PR. You and I were working on the same issue from different angles and came up with what I think are relatively equivalent solutions. Your "do eval printf '%s\\n' "${INCPATH}"" will expand the glob into a list of files for each wild carded element. The find does the same thing. I can't think of a case that your solution handles that the existing code doesn't. If you have one, I would love to see it, because I can't think of it.

I tried this 
```
ip-10-0-0-69:/home/ec2-user # cat /tmp/rsyslog.conf
##
## === When you're using remote logging, enable on-disk queues ===  
#
$IncludeConfig /etc/rsyslog.d/*.conf
$IncludeConfig /etc/rsyslog.d/*.xxxx
$IncludeConfig /etc/rsyslog.d/*.frule
$IncludeConfig /etc/rsyslog.d/


include(
   file="/path/to/include.conf"
   mode="optional"
)
```
and the result from your change is:
```
/etc/rsyslog.conf /etc/rsyslog.d/remote.conf /etc/rsyslog.d/*.xxxx /etc/rsyslog.d/NetworkManager.frule /etc/rsyslog.d/acpid.frule /etc/rsyslog.d/firewall.frule /etc/rsyslog.d/ /path/to/include.conf

```
the same input withe the current code gives:
```
/etc/rsyslog.conf /etc/rsyslog.d/remote.conf /etc/rsyslog.d/NetworkManager.frule /etc/rsyslog.d/acpid.frule /etc/rsyslog.d/firewall.frule /etc/rsyslog.d

```
The difference is that the files that don't exist were filtered out by the find. I created /etc/rsyslog.d/foo.tmp
and neither code base caught that file. I did not test nested includes. I know the exiting code does not handle those
and I can't see how your change would either.

brett060102 commented:
changing the find loop to:
```
RSYSLOG_FILES=()
for ENTRY in "${RSYSLOG_CONFIGS[@]}"
do
     echo "$(dirname "${ENTRY}")" " " "$(basename "${ENTRY}")"
     
     if [[ -d "${ENTRY}" ]]; then
        mapfile -t FINDOUT < <(find "${ENTRY}" -maxdepth 2 )
     else
        mapfile -t FINDOUT < <(find "$(dirname "${ENTRY}")" -maxdepth 2 -name "$(basename "${ENTRY}")")
     fi
     RSYSLOG_FILES+=("${FINDOUT[@]}")
done
```

fixes the directory problem, but does not address the nested include problem.

lonicerae commented:
Hi @brett060102 , I think if the simple way works we should keep it simple. In fact I believe this is a special usage of quoting, that's why I suggest using `eval` to show out the elements of the array.

As to the `rsyslog.conf` the pattern it can be `/etc/rsyslog.d/*/*/*.conf` etc. Hope this can make things clear.

brett060102 commented:
@lonicerae I am OK with that and I defer to @yuumasato and @matejak My concern is that we won't be handling:
$IncludeConfig /etc/rsyslog.d/

lonicerae commented:
@brett060102 Good catch! Thanks for pointing out this case. Please help to review the latest commit which I 'steal' part of your code. Thank you!

brett060102 commented:
Looks very good just a couple of things:
Based on: https://www.rsyslog.com/doc/v8-stable/configuration/global/options/rsconf1_includeconfig.html
$IncludeConfig /directory/ will include all regular files not starting with dot "."
So, I think:
readarray -t FINDOUT < <(find "${ENTRY}" -type f -name '*.conf')
should be 
readarray -t FINDOUT < <(find "${ENTRY}" -\( ! -regex '.*/\..*' \) -type f )

which seems to do the fight thing in my test case.

Also, I am not sure about the:
```
	else
		echo "Invalid include object: ${ENTRY}"
```
action. Since the following block in the config file:
```
include(
   file="/path/to/include.conf"
   mode="optional"
)
```
means include it if it exists, and ignore it if it doesn't then it is really not invalid. How about:
echo "Skipping ${ENTRY}: file not found"

brett060102 commented:
I updated my previous  comment find example to add maxdepth since based on my reading of the rysyslog.conf docs the direct include does not follow subdirs.

lonicerae commented:
That's strange.. The rsyslog behaves different against the doc. The logic is based on the testing results. I'll look into the code of rsyslog.

lonicerae commented:
I believe the feature:

```
Based on: https://www.rsyslog.com/doc/v8-stable/configuration/global/options/rsconf1_includeconfig.html
$IncludeConfig /directory/ will include all regular files not starting with dot "." and should not follow subdirs
```

Should have been removed long time ago by this commit: https://github.com/rsyslog/rsyslog/commit/5e2b03a31c312e6cd6a7f4cb0bca1f062668dc52 , maybe a bug for the document can be opened with rsyslog.

> Also, I am not sure about the:
> 
> ```
> 	else
> 		echo "Invalid include object: ${ENTRY}"
> ```
> 
> action. Since the following block in the config file:
> 
> ```
> include(
>    file="/path/to/include.conf"
>    mode="optional"
> )
> ```
> 
> means include it if it exists, and ignore it if it doesn't then it is really not invalid. How about: echo "Skipping ${ENTRY}: file not found"

Let's think about this pattern:

```
include(file="/dev/ttyS10" mode="optional")
include(
   file="/path/to/include.conf"
   mode="optional"
)
```

Apparently the `/dev/ttyS10` is not a regular file and should not be included. This is a defensive code, we should send out a meaningful message.

brett060102 commented:
@lonicerae  I have checked two systems Ubuntu 20.04 and SLE15 SP3, Both are running some version of rsyslogd version 8,
But the manpage version for rsyslog.conf in both cases is:
Version 7.2.0  22 October 2012

Man on rsyslogd shows.
Version 8.....    28 May 2014
what follows the 8 is vendor specific, but big point is that your guess about stale docs looks correct.
I did not actually play around with syslog. Are you stating that in your syslog testing that:
$IncludeConfig /etc/rsyslog.d/
Is no longer supported? Or if the check for regular files was removed?

You change addresses the issues that I hit and in cloud deployments. Which is my primary concern.

I am going to defer to the yuumasato and @matejak on:
echo "Invalid include object: ${ENTRY}"
We are creeping into more global topic that I don't feel competent to address. Should we be doing validation of the config files at remediation time or should that be being done at via the test oval. I have always thought of remediation as pretty quiet. Fix what we can and then retest.

lonicerae commented:
Hi @brett060102 ,

I compared relevant functions in last 10 years in rsyslog's repo and now indeed there're more changes..
```
I have checked two systems Ubuntu 20.04 and SLE15 SP3, Both are running some version of rsyslogd version 8,
But the manpage version for rsyslog.conf in both cases is:
Version 7.2.0 22 October 2012
```
It would be good if you could give me the links to the source of these versions, I can help to look into them. The upstream should be using the glob() to detect the wildcard since this commit:
 - https://github.com/rsyslog/rsyslog/commit/0f6300f1e6a038f37c406b2362218ea043def55b

In this case I need to update the `find` command accordingly. Please help to check the new commit https://github.com/ComplianceAsCode/content/pull/8726/commits/f4145979b2c653e961dc2fc1e7dac858b00e0610 and its test files, there I listed several different include patterns.

```
Are you stating that in your syslog testing that:
$IncludeConfig /etc/rsyslog.d/
Is no longer supported? Or if the check for regular files was removed?
```
No no, seems I was not explaining the issue clearly. The `$IncludeConfig /etc/rsyslog.d/` is absolutely supported. What we need is to carefully match the behavior of how rsyslog scanning and loading valid config files.

```
You change addresses the issues that I hit and in cloud deployments. Which is my primary concern.
```

May I know your use case? Could you share me an example?

```
We are creeping into more global topic that I don't feel competent to address. Should we be doing validation of the config files at remediation time or should that be being done at via the test oval. I have always thought of remediation as pretty quiet. Fix what we can and then retest.
```
No worries, as I work and test rsyslog a lot in my job so I may see lots of strange issues. :stuck_out_tongue: I'm sorry that I don't now much of the framework design of this `ComplianceAsCode/content` project. I also would like to learn more about this project, that's interesting!

lonicerae commented:
> Should have been removed long time ago by this commit: https://github.com/rsyslog/rsyslog/commit/5e2b03a31c312e6cd6a7f4cb0bca1f062668dc52 ,
> maybe a bug for the document can be opened with rsyslog.

The `dot file` handling has been handled by the above code after rsyslog refactoring. Now not only the hidden files `.xxxx` but also the files in hidden directories `.xxxx/yyyy` should be ignored if we use a directory in `file` option with `include()` or `IncludeConfig` directives.

lonicerae commented:
The new commit fixed an incorrect parsing of rules when extracting the log file path from config files. It's due to the incompatibility of the old code with the new syntax of the rsyslog directives. As I mentioned in the old PR, it would be good to write parsers for parsing the rsyslog settings in the long term.

brett060102 commented:
@lonicerae I was hitting the 
$IncludeConfig /etc/rsyslog.d/*.conf
Specifically:
/etc/rsyslog.d/21-cloudinit.conf
I noticed it because 
/var/log/cloud-init.log
Did not have the right privs, this problem was detected by the oval, after running remediation the problem still existed, because the we were not remediating files listed in the in the $IncludeConfig /etc/rsyslog.d/*.conf files.

In normal operation a rule is only remediated if, the test done by the rules fails. So, normal sequence would be:
1: rest rule
2: if test == fail => remediate and retest
3: if retest == fail; flag remediation as failed.
So, if the oval does not fail, remediation is not run. Which is why the original problem was around so long.

sample command would be:
$ sudo /usr/bin/oscap xccdf eval --remediate --profile xccdf_org.ssgproject.content_profile_cis /usr/share/xml/scap/ssg/content/ssg-sle15-ds-1.2.xml

It looks like you have rsyslog expertise, which I don't have. So, I just want to make sure I understand this correctly:

1: A directory include in rsyslog will exclude and hidden files, but include all others traverse any directories.
2: wildcards are permitted in not only for file names, but also in file paths.

If that is correct:
1: it does not match what I find in the documentation. Wildcards in paths are not documented and  I think the most recent 
    documentation states: "If all regular files in the /etc/rsyslog.d directory are included, then files starting with ”.” are ignored"
    Including sudbirs is not explicitly stated, and the use of "regular files" I interpret as would pass the "-f FILENAME" test which
  is not the case for directories.
2: I think it opens a potential security hole. Since someone could add a conf file somewhere in the that path chain that  would cause privs to be updated if/when remediation is run. 

Maybe @yuumasato or @matejak can comment here. My thoughts are:
1: from a security perspective wildscards are not a good idea. Since, $IncludeConfig /etc/rsyslog.d/*.conf is well documented and is supported by the oval. So, I will live with that one, but the wildcards in the path spec just feel pretty dangerous.
2: We should only remediated issues that are detected by the associated test oval.  In looking over the oval, I don't think that it would handle the directory include and I am not sure about the wildcard in the path. It might handle that one.
3: You do seem to cover the testing scenarios in your test code. While I think a single global pass test would be OK, I think the failures case should be distinct, because while any single failure will cause a pass test to fail. Any single failure will also cause a fail test to pass. So, I think that each the failure path need to be tested separately.

lonicerae commented:
Hi @brett060102 , thanks for your comments, please check my reply inline.

```
> I was hitting the
> $IncludeConfig /etc/rsyslog.d/.conf
> Specifically:
> /etc/rsyslog.d/21-cloudinit.conf
> I noticed it because
> /var/log/cloud-init.log
> Did not have the right privs, this problem was detected by the oval, after running remediation the problem still existed,
> because the we were not remediating files listed in the in the $IncludeConfig /etc/rsyslog.d/.conf files.
```
Could you show me the contents of your `/etc/rsyslog.d/.conf` and `/etc/rsyslog.d/21-cloudinit.conf`? I try to file the `/etc/rsyslog.d/.conf` with the following:

```
local4.*                                                /var/log/local4.log
```

And the script can show the correct log file `/var/log/local4.log` for remediation.

```
> In normal operation a rule is only remediated if, the test done by the rules fails. So, normal sequence would be:
> 1: rest rule
> 2: if test == fail => remediate and retest
> 3: if retest == fail; flag remediation as failed.
> So, if the oval does not fail, remediation is not run. Which is why the original problem was around so long.
>
> sample command would be:
> $ sudo /usr/bin/oscap xccdf eval --remediate --profile xccdf_org.ssgproject.content_profile_cis /usr/share/xml/scap/ssg/content/ssg-sle15-ds-1.2.xml
```
Thanks for sharing the test procedure! I'll use it for testing in the future.

```
> It looks like you have rsyslog expertise, which I don't have. So, I just want to make sure I understand this correctly:
>
> 1: A directory include in rsyslog will exclude and hidden files, but include all others traverse any directories.
```
Based on the rsyslog's code, it depends on how you write your `include config` diretive.
1. If we have:
```
$IncludeConfig /etc/rsyslog.d/.conf
```
then the content of `/etc/rsyslog.d/.conf` will be read and parsed.

2. If we only include the directory instead of a specific full path:
```
$IncludeConfig /etc/rsyslog.d/
``` 
then the `/etc/rsyslog.d/.conf` will not be read and will not be parsed.
So I don't have much confidence in current code in master.

```
> 2: wildcards are permitted in not only for file names, but also in file paths.
```
Yes. In fact that's a pattern to be expanded.

```
> If that is correct:
> 1: it does not match what I find in the documentation. Wildcards in paths are not documented and I think the most recent
> documentation states: "If all regular files in the /etc/rsyslog.d directory are included, then files starting with ”.” are ignored"
> Including sudbirs is not explicitly stated, and the use of "regular files" I interpret as would pass the "-f FILENAME" test which
> is not the case for directories.
```
Hmm.. That's why I think the current code can not cover all cases of the `include config`.

```
> 2: I think it opens a potential security hole. Since someone could add a conf file somewhere in the that path chain that
> would cause privs to be updated if/when remediation is run.
```
My friend and I had thought about this when I submitted the old PR. We don't know if the framework of this project will filter out these paths.

```
> Maybe @yuumasato or @matejak can comment here. My thoughts are:
> 1: from a security perspective wildscards are not a good idea. Since, $IncludeConfig /etc/rsyslog.d/*.conf is well documented
> and is supported by the oval. So, I will live with that one, but the wildcards in the path spec just feel pretty dangerous.
```
I think as long as you are processing the `include config` directive, you have been facing a lot of things, no only the wildcard.

```
> 2: We should only remediated issues that are detected by the associated test oval. In looking over the oval, I don't think that
> it would handle the directory include and I am not sure about the wildcard in the path. It might handle that one.
> 3: You do seem to cover the testing scenarios in your test code. While I think a single global pass test would be OK, I think
> the failures case should be distinct, because while any single failure will cause a pass test to fail. Any single failure will also
> cause a fail test to pass. So, I think that each the failure path need to be tested separately.
```
Yeah, maybe. A thought comes up when I was writing the test code is that maybe the framework could provide a `pytest` like template for testing. Sorry that I need to focus more on my work in the next few weeks and may not have much time on this. It doesn't matter if this PR is not accepted. I learned a lot from this.

yuumasato commented on linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:
```diff
@@ -5,28 +5,39 @@
-# So, loop over the entries in RSYSLOG_CONFIGS and use find to get the list of included files.
-RSYSLOG_FILES=()
+# Array to hold all rsyslog config files
+declare -a RSYSLOG_CONFIG_FILES
```
```suggestion
RSYSLOG_FILES=()
```

yuumasato commented on linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:
```diff
@@ -5,28 +5,39 @@
-RSYSLOG_CONFIGS=()
-RSYSLOG_CONFIGS=("${RSYSLOG_ETC_CONFIG}" "${RSYSLOG_INCLUDE_CONFIG[@]}" "${RSYSLOG_INCLUDE[@]}")
+# Array to hold all rsyslog config entries
+declare -a RSYSLOG_CONFIGS
```
```suggestion
RSYSLOG_CONFIGS=()
```
Please keep the `RSYSLOG_CONFIGS=()` syntax.
This ensures that the array is clean and empty.

yuumasato commented on linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:
```diff
@@ -5,28 +5,39 @@
+	then
+		RSYSLOG_CONFIG_FILES+=("${ENTRY}")
+	else
+		echo "Invalid include object: ${ENTRY}"
```
I have checked the OVAL code and I think will fail if an explicit include (not regex) to a file that doesn't exist is defined, even the include has `mode=optional`.
But the best way to know the behaviour for sure is to have a test scenario, ;)
It would be great if you could add it, a test where ` include(file="${RSYSLOG_TEST_DIR}/missing_file.conf")` is specified but the file doesn't exist.

yuumasato commented on linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:
```diff
@@ -5,28 +5,39 @@
+	# However, files in hidden sub-directories or hidden files will be ignored.
+	if [ -d "${ENTRY}" ]
+	then
+		readarray -t FINDOUT < <(find "${ENTRY}" -not -path '*/.*' -type f)
```
By reading the OVAL check I don't think it will ignore hidden files.
A pass test where files with correct permissions are set, and only the hidden files or files in hidden directories have wrong permissions will clarify this.

yuumasato reviewed (commented):
@lonicerae @brett060102 Hi, sorry for my delay and  thank you for the great technical discussion.

I tried to provide input regarding behaviour of the check, but I think the best way to assert and ensure how it works is via test scenarios, :)

brett060102 commented:
@lonicerae  Sorry for the delay. My /etc/rsyslog.d/21-cloudinit.conf has 

```
ec2-user@ip-10-0-0-14:~> cat /etc/rsyslog.d/21-cloudinit.conf 
# Log cloudinit generated log messages to file
:syslogtag, isequal, "[CLOUDINIT]" /var/log/cloud-init.log

# comment out the following line to allow CLOUDINIT messages through.
# Doing so means you'll also get CLOUDINIT messages in /var/log/syslog
& stop
```

yuumasato commented:
BTW, I can provide help or code if OVAL changes are needed.

yuumasato commented:
Closing as #9326 got merged.

yuumasato commented:
Thnank you @lonicerae for the patch
